### ISSUE TITLE
Bitbucket credentials not persisting

### ISSUE THREAD

**Issue by bahia0019** [afragen/github-updater#662]
I am using your plugin on a staging site currently: https://staging.thebirdthebear.com/
When it works, it's amazing, but I'm having serious issues with it.

When I log in to check for updates I've made, I get this screen even though I've installed a theme and connected to Bitbucket already, and set it up properly:
<img width="873" alt="screen shot 2018-03-26 at 4 22 19 pm" src="https://user-images.githubusercontent.com/1586037/37938414-94c69852-3112-11e8-95ec-a28fe627960e.png">

I found that if I click Save, the page will refresh, and I'll get this page:
<img width="858" alt="screen shot 2018-03-26 at 4 22 34 pm" src="https://user-images.githubusercontent.com/1586037/37938449-c2f56230-3112-11e8-9981-058ebc796239.png">

I'll click over to the Bitbucket tab and my theme shows, but I have to re-enter everything, and check that it's Private repo again:
<img width="952" alt="screen shot 2018-03-26 at 4 22 47 pm" src="https://user-images.githubusercontent.com/1586037/37938477-e41cff36-3112-11e8-91fe-c83a9b341d87.png">

I do:
<img width="938" alt="screen shot 2018-03-26 at 4 23 13 pm" src="https://user-images.githubusercontent.com/1586037/37938480-eaf6a438-3112-11e8-9bda-bd88309c12d4.png">

Hit enter, and this screen displays:
<img width="946" alt="screen shot 2018-03-26 at 4 23 24 pm" src="https://user-images.githubusercontent.com/1586037/37938503-046e91be-3113-11e8-8a96-95bcefa5ff80.png">

At this stage, the new theme changes on Bitbucket haven't been pulled, and I'm pretty much where I started.
The last couple changes, I've just uploaded via FTP since this hasn't worked.

Happy to provide any information that would help in resolving this. 
If you need error logs, or access to the staging site, I'm happy to provide.

**Comment by afragen** [comment]
What branch and version of the plugin are you using?

**Comment by bahia0019** [comment]
7.5.0.12
Not sure the branch. I grabbed it from the main repo.

**Comment by afragen** [comment]
Please use the branch switch to reinstall the latest `develop` branch and see if the issue persists.

**Comment by bahia0019** [comment]
Looks like that might have done the trick. I'll push a couple more versions and keep you posted.
Thanks!

**Comment by bahia0019** [comment]
Hmmm... Not working.
The theme name and the checkbox DO stay at least. But I pushed a new version earlier tonight, hit refresh, nothing. Waited awhile to see if that would help, nothing.

Any chance it's due to using a subdomain for staging? Wild guess.

**Comment by afragen** [comment]
Can you check out the wiki page on background processing and possibly use the referenced filter hook to bypass WP-Cron processing. 

https://github.com/afragen/github-updater/wiki/Background-Processing

**Comment by bahia0019** [comment]
Thanks! 
I've fixed this, and will push some stuff later today and update you.
Good catch.

**Comment by bahia0019** [comment]
Seems to have gone back to the buggy behavior I submitted about.
Everything is missing.

I gotta finish this project. So I'm switching back to FTP. 
If you want access to dig further, I'm happy to provide.

**Comment by afragen** [comment]
I’m sorry it’s not working for you. Do what you need to do for your project. I hope when you have some free time we can figure out the issue.

**Comment by afragen** [comment]
Is it possible to give `afragen` `read` access to the repo so that I can test on my end?

**Comment by FORTE-WP** [comment]
Hi Andy, thanks for this great plugin. It has helped me for some time now. But I recently seem to have the same problem as Willam. On refresh the Bitbucket tab disappears. And updates are not carried out anymore.  Greetings, Frank

**Comment by FORTE-WP** [comment]
I've just done some testing. The problem started in release 7.5.0. My website with 7.4.4. is working fine.

**Comment by afragen** [comment]
I just released 7.6.0. Is this issue still present?

**Comment by FORTE-WP** [comment]
I'm sorry but it is not resolved in that version. Before I posted I had
already updated.

Op ma 9 apr. 2018 15:07 schreef Andy Fragen <notifications@github.com>:

**Comment by afragen** [comment]
@FORTE-WP can you start a new issue and describe precisely what you see and how to replicate it?

**Comment by afragen** [comment]
I think I fixed this issue in https://github.com/afragen/github-updater/commit/8c2fdd30ff39ddb88d4173e7c2dfcb0d5a6cc28d

**Comment by afragen** [comment]
@bahia0019 can you let me know if the latest release fixes this issue for you? Thanks.

**Comment by bahia0019** [comment]
@afragen I'll be doing some work on my project tomorrow. I'll push up a few times and give it a good test and report back.

**Comment by afragen** [comment]
Sounds great!

**Comment by bahia0019** [comment]
I just pushed a new commit, and was going to update it in the Admin, but didn't see Bitbucket again. 
Clicked Save, and this is the screen I get now.
So... Still having some issues.
<img width="666" alt="screen shot 2018-04-19 at 11 56 25 am" src="https://user-images.githubusercontent.com/1586037/39012253-cb20bef0-43c8-11e8-8bf5-b400bfc567a6.png">

**Comment by afragen** [comment]
What does the Bitbucket tab look like?

**Comment by bahia0019** [comment]
It was empty. But I installed the theme yesterday. 
<img width="802" alt="screen shot 2018-04-19 at 12 32 00 pm" src="https://user-images.githubusercontent.com/1586037/39013922-bc288b4e-43cd-11e8-9cc8-b664e3ba7738.png">

**Comment by afragen** [comment]
This is odd. You are using version 7.6.1 or later and the theme is installed?

Can you check the contents of `$installed_apis`?

**Comment by afragen** [comment]
Maybe set the scope of `$installed_apis` from protected to public in `Base` and see if that does anything.

**Comment by afragen** [comment]
Can I ask what version of PHP you’re using?

**Comment by bahia0019** [comment]
Just checked PHP. It's using 7.0
Not sure what static::$installed_apis is.

**Comment by afragen** [comment]
So when Settings is called it extends class Base. The constructor of Base runs  `set_installed_apis()` and stores the results into `Base::$installed_apis`. Since Settings extends Base this static should be set. To load all the subtabs `Settings::settings_sub_tabs()` tests and loads each `static::$installed_apis` to load that sub tab.

**Comment by bahia0019** [comment]
Sorry,
Not really sure what I need to do here. 
I'm happy to provide access to the staging server, and the WordPress install if you want to run tests. If I get out of the way, you might find answers easier.

Let me know.

**Comment by afragen** [comment]
As much as I don’t like playing on someone else’s server, I’m game. Please make sure you have backups I don’t want to inadvertently break something. Also, my laptop is out of commission for a few more days so I’ll see what I can see on my iPad. Sorry not ideal.

**Comment by afragen** [comment]
andy at thefragens dot com

**Comment by afragen** [comment]
I was able to log in. The Settings are empty, missing both a GitHub Access Token and Bitbucket credentials.

**Comment by afragen** [comment]
What did you do to have the GitHub and Bitbucket tabs not show up?

**Comment by bahia0019** [comment]
Yeah, they were empty when I came back to update. But I did install the theme.
I don't know what I did. I installed the theme. Pushed a change. Everything was fine.
I attempted to do the same today, and the Github and Bitbucket tabs were gone.

**Comment by afragen** [comment]
So, I re-installed GHU. You’re currently on the `master` branch. Sometimes there are issues if you don’t correctly rename GHU before installing. Unfortunately I’m not certain where those issues my be. I know one of them is that the `Settings` link doesn’t show in the plugins.php page next to the `Deactivate` link.

**Comment by bahia0019** [comment]
OK, if you need to test more, I'm happy to offer up the staging site.
So, at this point, I just re-add my credentials and keep the "Enable branch switching" checked?

I'll be pushing a couple things tonight. I can keep you updated. Let me know if you need access to anything else.

Thanks!

**Comment by afragen** [comment]
Just enter your credentials. I deactivated branch switching but you can turn it on if you wish.

**Comment by afragen** [comment]
Unfortunately what this means is I didn’t see anything not working. Let me know if you see it again and I’ll log in to take a look.

**Comment by afragen** [comment]
So I now see this error on your site. I’m still without a computer but it gives me a place to start.

**Comment by bahia0019** [comment]
@afragen Hey Andy. I wanted to get back to you on this. I just finished up my project, and pushed a lot in the process. It's super solid. I had absolutely no problems after that update.

Thanks for working on that so hard. This definitely makes things SO much easier. I appreciate your efforts with this plugin!
